caas/kubernetes/provider/application/application.go

```diff
@@ -0,0 +1,791 @@
+	switch a.deploymentType {
+	case caas.DeploymentStateful:
+		informer = factory.Apps().V1().StatefulSets().Informer()
+	case caas.DeploymentStateless:
+		informer = factory.Apps().V1().Deployments().Informer()
+	case caas.DeploymentDaemon:
+		informer = factory.Apps().V1().DaemonSets().Informer()
+	default:
+		return nil, errors.NotSupportedf("unknown deployment type")
+	}
+	return a.newWatcher(informer, a.name, a.clock)
+}
+
+func (a *app) WatchReplicas() (watcher.NotifyWatcher, error) {
+	factory := informers.NewSharedInformerFactoryWithOptions(a.client, 0,
+		informers.WithNamespace(a.namespace),
+		informers.WithTweakListOptions(func(o *metav1.ListOptions) {
+			o.LabelSelector = a.labelSelector()
+		}),
+	)
+	return a.newWatcher(factory.Core().V1().Pods().Informer(), a.name, a.clock)
+}
+
+func (a *app) State() (caas.ApplicationState, error) {
+	state := caas.ApplicationState{}
+	switch a.deploymentType {
+	case caas.DeploymentStateful:
+		ss := resources.NewStatefulSet(a.name, a.namespace, nil)
+		err := ss.Get(context.Background(), a.client)
+		if err != nil {
+			return caas.ApplicationState{}, errors.Trace(err)
+		}
+		if ss.Spec.Replicas == nil {
+			return caas.ApplicationState{}, errors.Errorf("missing replicas")
+		}
+		state.DesiredReplicas = int(*ss.Spec.Replicas)
+	case caas.DeploymentStateless:
+		d := resources.NewDeployment(a.name, a.namespace, nil)
+		err := d.Get(context.Background(), a.client)
+		if err != nil {
+			return caas.ApplicationState{}, errors.Trace(err)
+		}
+		if d.Spec.Replicas == nil {
+			return caas.ApplicationState{}, errors.Errorf("missing replicas")
+		}
+		state.DesiredReplicas = int(*d.Spec.Replicas)
+	case caas.DeploymentDaemon:
+		d := resources.NewDaemonSet(a.name, a.namespace, nil)
+		err := d.Get(context.Background(), a.client)
+		if err != nil {
+			return caas.ApplicationState{}, errors.Trace(err)
+		}
+		state.DesiredReplicas = int(d.Status.DesiredNumberScheduled)
+	default:
+		return caas.ApplicationState{}, errors.NotSupportedf("unknown deployment type")
+	}
+	next := ""
+	for {
+		res, err := a.client.CoreV1().Pods(a.namespace).List(context.Background(), metav1.ListOptions{
+			LabelSelector: a.labelSelector(),
+			Continue:      next,
+		})
+		if err != nil {
+			return caas.ApplicationState{}, errors.Trace(err)
+		}
+		for _, pod := range res.Items {
+			state.Replicas = append(state.Replicas, pod.Name)
+		}
+		if res.RemainingItemCount == nil || *res.RemainingItemCount == 0 {
+			break
+		}
+		next = res.Continue
+	}
+	sort.Strings(state.Replicas)
+	return state, nil
+}
+
+// applicationPodSpec returns a PodSpec for the application pod
+// of the specified application.
+func (a *app) applicationPodSpec(config caas.ApplicationConfig) (*corev1.PodSpec, error) {
+	appSecret := a.secretName()
+
+	if config.Charm.Meta().Deployment == nil {
+		return nil, errors.NotValidf("charm missing deployment")
+	}
+	containerImageName := config.Charm.Meta().Deployment.ContainerImageName
```

Comment:
These attributes should be part of the juju model, not accessed directly off the charm

Reply:
This again will depend on the charm metadata in the container section. Will likely be resources.